Use S-LoRA for inference

corbt opened the pull request:
This PR switches us to the S-LoRA branch of vLLM and starts serving future fine-tunes via S-LoRA, instead of by serving the fully-merged model.

It does however maintain backwards compatibility. After training a model, we still persist the merged model to disk, and we could switch to serving it via the old process just by setting the fine-tune's pipeline version to 2 instead of 3.

There is an (intentional) regression here: for new pipeline v3 models trained in prod, we'll no longer be able to automatically serve them in development. I'll need to write a script that we can call to explicitly copy production models into development when necessary for testing.

corbt committed:
Use S-LoRA for inference

This PR switches us to the S-LoRA branch of vLLM and starts serving future fine-tunes via S-LoRA, instead of by serving the fully-merged model.

It does however maintain backwards compatibility. After training a model, we still persist the merged model to disk, and we could switch to serving it via the old process just by setting the fine-tune's pipeline version to 2 instead of 3.

There is an (intentional) regression here: for new pipeline v3 models trained in prod, we'll no longer be able to automatically serve them in development. I'll need to write a script that we can call to explicitly copy production models into development when necessary for testing.

arcticfly commented on app/package.json:
```diff
@@ -20,13 +20,14 @@
     "repl": "tsx scripts/repl.ts",
     "repl:watch": "tsx --watch scripts/repl.ts",
-    "dev:modal": "cd ../trainer && MODAL_ENVIRONMENT='dev' concurrently 'poetry run modal serve src/inference_server/main.py' 'poetry run modal serve src/trainer/main.py' --kill-others"
+    "dev:modal": "cd ../trainer && MODAL_ENVIRONMENT='dev' concurrently 'poetry run modal serve src/lora_inference/main.py' 'poetry run modal serve src/trainer/main.py' --kill-others"
```
Does `lora_inference/main.py` also set up the serving of existing merged models?

corbt replied:
No, those will continue to be served from the endpoint defined in inference_server/main.py

arcticfly commented on app/src/modelProviders/fine-tuned/getCompletion.ts:
```diff
@@ -51,7 +45,14 @@ export async function getCompletion2(
   } else {
-    return getModalCompletion(fineTune, prunedInput);
+    switch (fineTune.pipelineVersion) {
+      case 1:
```
Could be worth commenting what each of these pipeline versions is. I'm starting to lose track.

corbt replied:
Added comments now

arcticfly commented on trainer/src/lora_inference/main.py:
```diff
@@ -0,0 +1,165 @@
+
+@stub.function(
+    allow_concurrent_inputs=500,
+    timeout=1 * 60 * 60,
```
Is the timeout now an hour?

arcticfly commented on trainer/src/shared.py:
```diff
@@ -47,3 +49,8 @@ def upload_directory_to_s3(local_directory, destination, bucket):
             future.result()
+
+
+def require_auth(auth_credentials=Depends(HTTPBearer())):
```
👏

arcticfly reviewed (approved):
Very excited to try this out!x

corbt committed:
- Keep on pipeline 2 by default, make pipeline 3 experimental
 - Documentation of pipelines
 - Store LoRAs by their database uuid, not by their huggingface id

corbt committed:
use more router functions with lora pipeline
